File: pwiz/analysis/demux/OverlapDemultiplexer.cpp
//
// $Id$
//
//
// Original author: Jarrett Egertson <jegertso .@. uw.edu>
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#include "OverlapDemultiplexer.hpp"
#include "SpectrumPeakExtractor.hpp"
#include "IInterpolation.hpp"
#include "DemuxHelpers.hpp"
#include "CubicHermiteSpline.hpp"
#include <boost/make_shared.hpp>

namespace pwiz {
namespace analysis {
    using namespace std;
    using namespace DemuxTypes;
    using namespace msdata;

    OverlapDemultiplexer::OverlapDemultiplexer(Params p) :
        params_(p)
    {
        overlapRegionsInApprox_ = 7;
        cyclesInBlock_ = 3;
    }

    OverlapDemultiplexer::~OverlapDemultiplexer()
    {
    }

    void OverlapDemultiplexer::Initialize(SpectrumList_const_ptr sl, IPrecursorMaskCodec::const_ptr pmc)
    {
        assert(sl);
        assert(pmc);
        if (!sl || !pmc)
            return;
        sl_ = sl;
        pmc_ = pmc;
    }

    void OverlapDemultiplexer::BuildDeconvBlock(size_t index, const vector<size_t>& muxIndices, MatrixPtr& masks, MatrixPtr& signal) const
    {
        if (!sl_ || !pmc_)
            throw runtime_error("BuildDeconvBlock() Null pointer to SpectrumList and/or IPrecursorMaskCodec. OverlapDemultiplexer may not have been initialized.");

        // get the list of peaks to demultiplex
        Spectrum_const_ptr deconvSpectrum = sl_->spectrum(index, true);
        BinaryDataArrayPtr mzsToDemux = deconvSpectrum->getMZArray();
        SpectrumPeakExtractor peakExtractor(mzsToDemux->data, params_.massError);

        // initialize mask and intensities matrices.
        size_t numMuxSpectra = overlapRegionsInApprox_; // m
        size_t numDemuxSpectra = overlapRegionsInApprox_; // n
        size_t numTransitions = mzsToDemux->data.size(); // k
        masks = boost::make_shared<MatrixType>(numMuxSpectra, numDemuxSpectra);
        signal = boost::make_shared<MatrixType>(numMuxSpectra, numTransitions);
        
        // Find the scans nearest to the deconvolution windows of the spectrum of interest
        vector<size_t> deconvIndices;
        pmc_->SpectrumToIndices(deconvSpectrum, deconvIndices);
        double sum = 0.0;
        for (auto i : deconvIndices) { sum += i; }
        auto centerOfDeconvIndices = sum / deconvIndices.size();

        // Find the range of deconv window indices that will be used to extract the mask for this spectrum
        auto idealLowerMzBound = lround(centerOfDeconvIndices - overlapRegionsInApprox_ / 2.0);
        auto lowerMZBound = max(idealLowerMzBound, 0l);
        lowerMZBound = min(lowerMZBound, static_cast<long>(pmc_->GetNumDemuxWindows() - overlapRegionsInApprox_));

        vector<pair<double, size_t> > demuxWindowDistances;
        auto specPerCycle = pmc_->GetSpectraPerCycle();
        Spectrum_const_ptr offsetSpectrum;
        vector<size_t> offsetIndices;
        for (auto scanIndex = muxIndices.begin(); scanIndex < muxIndices.end(); ++scanIndex)
        {
            offsetSpectrum = sl_->spectrum(*scanIndex, true);
            pmc_->SpectrumToIndices(offsetSpectrum, offsetIndices);
            double offsetSum = 0.0;
            for (auto i : offsetIndices) { offsetSum += i; }
            auto centerOfOffsetIndices = offsetSum / offsetIndices.size();
            auto distance = centerOfOffsetIndices - centerOfDeconvIndices;
            demuxWindowDistances.push_back(pair<double, size_t>(distance, *scanIndex));
        }

        // Sort by distances from spectrum of interest in m/z space
        sort(demuxWindowDistances.begin(), demuxWindowDistances.end(), [](const pair<double, size_t>& left, const pair<double, size_t>& right)
        {
            return abs(left.first) < abs(right.first) && abs(abs(left.first) - abs(right.first)) > 10e-4;
        });

        // Choose the regions near enough to the spectrum demux windows of interest to be included
        vector<pair<double, size_t>> bestMaskAverages;
        bestMaskAverages.reserve(overlapRegionsInApprox_);
        copy(demuxWindowDistances.begin(), demuxWindowDistances.begin() + overlapRegionsInApprox_, back_inserter(bestMaskAverages));

        sort(bestMaskAverages.begin(), bestMaskAverages.end(), [](const pair<double, size_t>& left, const pair<double, size_t>& right)
        {
            return left.first < right.first && abs(left.first - right.first) > 10e-4;
        });

        // Get the indices of the chosen regions
        vector<size_t> scansInDeconv;
        scansInDeconv.reserve(bestMaskAverages.size());
        transform(bestMaskAverages.begin(), bestMaskAverages.end(), back_inserter(scansInDeconv), [](const pair<double, size_t> &p)
        {
            return p.second;
        });

        // Fill masks
        for (size_t matrixRow = 0; matrixRow < scansInDeconv.size(); ++matrixRow)
        {
            size_t currentIndex = scansInDeconv[matrixRow];
            Spectrum_const_ptr muxSpectrum = sl_->spectrum(currentIndex, true);
            auto fullMaskRow = pmc_->GetMask(muxSpectrum); /// GetMask could be made more efficient to only return a specified subset of masks
            masks->row(matrixRow) = fullMaskRow.segment(lowerMZBound, numDemuxSpectra);
        }

        // Fill signal
        if (params_.interpolateRetentionTime)
        {
        // Get retention time for scan to be deconvolved
        double deconvStartTime;
        if (!TryGetStartTime(*deconvSpectrum, deconvStartTime))
            throw runtime_error("BuildDeconvBlock() Tried to process an MS2 scan without retention times written.");

        // Cache the data that will be used for interpolation
        vector<MatrixPtr> binnedIntensitiesCache;
        vector<boost::shared_ptr<VectorXd> > scanTimesCache;

        // Loop through the scans and cache intensity and time data for interpolation
        for (size_t matrixRow = 0; matrixRow < scansInDeconv.size(); ++matrixRow)
        {
            binnedIntensitiesCache.push_back(boost::make_shared<MatrixType>(cyclesInBlock_, numTransitions));
            scanTimesCache.push_back(boost::make_shared<VectorXd>(cyclesInBlock_));

            // Find scans around the spectrum to use for interpolation
            auto scan = scansInDeconv[matrixRow];
            vector<size_t> interpolationSpectraIndices;
            if (!FindNearbySpectra(interpolationSpectraIndices, sl_, scan, cyclesInBlock_, specPerCycle))
                throw runtime_error("BuildDeconvBlock() Not enough spectra to interpolate for the overlap.");

            // Extract retention times and peak intensities for scan
            for (size_t i = 0; i < interpolationSpectraIndices.size(); ++i)
            {
                auto spectrumIndex = interpolationSpectraIndices.at(i);
                Spectrum_const_ptr currentSpectrum = sl_->spectrum(spectrumIndex, true);
                
                double startTime;
                if (!TryGetStartTime(*currentSpectrum, startTime))
                    throw runtime_error("BuildDeconvBlock() Tried to process an MS2 scan without retention times written.");

                // Record scan time
                (*scanTimesCache.back())[i] = startTime;

                // Bin and record transitions
                peakExtractor(currentSpectrum, *binnedIntensitiesCache.back(), i);
            }
        }
        // Perform interpolation
        // TODO parallelize this
        for (size_t matrixRow = 0; matrixRow < overlapRegionsInApprox_; ++matrixRow)
        {
            InterpolateMuxRegion(signal->row(matrixRow), deconvStartTime, *binnedIntensitiesCache.at(matrixRow), *scanTimesCache.at(matrixRow));
        }
        }
        else
        {
            int specPerCycle = pmc_->GetSpectraPerCycle();
            for (int matrixRow = 0; matrixRow < scansInDeconv.size(); ++matrixRow)
            {
                size_t currentIndex = scansInDeconv[matrixRow];
                Spectrum_const_ptr s = sl_->spectrum(currentIndex, true);
                DemuxScalar weight = 1.0;
                if (params_.applyWeighting)
                {
                    /* This method of weighting tries to minimize the added variance from changing intensity during
                    * chromatogram elution. Given a point on the elution peak, one can model the change in intensity as a
                    * function of difference in retention time. By modeling this function for all possible points on the
                    * peak it is possible to find a best fit weighting scheme that minimizes intensity variance. To
                    * describe the model itself in more detail, ScanDiff contains the difference in retention time and
                    * 5 / specPerCycle contains information about the typical peak width (or standard deviation).
                    */
                    int scanDiff = int(index) - int(currentIndex);
                    weight = 1.0 / (1.0 + pow(5.0 * scanDiff / specPerCycle, 2));
                }
                peakExtractor(s, *signal, matrixRow, weight);
            }
        }

        // Cache the indices for the spectrum
        spectrumIndices_.clear();
        for (auto demuxIndex : deconvIndices)
        {
            assert(demuxIndex >= lowerMZBound);
            spectrumIndices_.push_back(demuxIndex - lowerMZBound);
        }
    }

    void OverlapDemultiplexer::GetMatrixBlockIndices(size_t indexToDemux, std::vector<size_t>& muxIndices, double demuxBlockExtra) const
    {
        demuxBlockExtra = max(0.0, demuxBlockExtra);
        auto numSpectraToFind = pmc_->GetSpectraPerCycle() + size_t(round(demuxBlockExtra * pmc_->GetSpectraPerCycle()));
        if (!FindNearbySpectra(muxIndices, sl_, indexToDemux, numSpectraToFind))
            throw runtime_error("GetMatrixBlockIndices() Not enough spectra to demultiplex this block");
    }

    const std::vector<size_t>& OverlapDemultiplexer::SpectrumIndices() const
    {
        return spectrumIndices_;
    }

    void OverlapDemultiplexer::InterpolateMuxRegion(Ref<MatrixXd, 0, Stride<Dynamic, Dynamic> > interpolatedIntensities, double timeToInterpolate,
        Eigen::Ref<const Eigen::MatrixXd> intensities, Ref<const VectorXd> scanTimes)
    {
        if (interpolatedIntensities.cols() != 1 && interpolatedIntensities.rows() != 1)
            throw runtime_error("InterpolateMuxRegion() Output block is not a vector type");

        if (interpolatedIntensities.size() != intensities.cols())
            throw runtime_error("InterpolateMuxRegion() Output block does not have the expected size");

        //TODO parallelize this
        auto numTransitions = static_cast<size_t>(intensities.cols());
        for (size_t transition = 0; transition < numTransitions; ++transition)
        {
            auto intensitiesBlock = intensities.col(transition);            
            interpolatedIntensities(0, transition) = max(0.0, InterpolateMatrix(timeToInterpolate, scanTimes, intensitiesBlock));
        }
    }

    double OverlapDemultiplexer::InterpolateMatrix(double pointToInterpolate, Eigen::Ref<const Eigen::VectorXd> points, Eigen::Ref<const Eigen::VectorXd> values)
    {
        boost::shared_ptr<IInterpolation> interpolator = boost::make_shared<CubicHermiteSpline>(points, values);
        return interpolator->Interpolate(pointToInterpolate);
    }
} // namespace analysis
} // namespace pwiz

File: pwiz-sharp/pwiz/src/Analysis/Demux/OverlapDemultiplexer.cs
using MathNet.Numerics.Interpolation;
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Double;
using Pwiz.Data.Common.Cv;
using Pwiz.Data.MsData.Spectra;
using Pwiz.Util.Chemistry;

namespace Pwiz.Analysis.Demux;

/// <summary>
/// Demultiplexer for overlap-style DIA experiments. Port of cpp's <c>OverlapDemultiplexer</c>.
/// Builds a smaller, locally-relevant deconv block (only the overlap regions adjacent in m/z to
/// the spectrum being demultiplexed) and optionally interpolates each block row to the
/// retention time of the demuxed spectrum via cubic Hermite splines (MathNet's
/// <c>CubicSpline.InterpolateHermite</c>).
/// </summary>
public sealed class OverlapDemultiplexer : IDemultiplexer
{
    /// <summary>Tunable parameters.</summary>
    public sealed class Params
    {
        /// <summary>Mass tolerance for MS/MS peak extraction (cpp default 10 ppm).</summary>
        public MZTolerance MassError { get; init; } = new(10.0, MZToleranceUnits.Ppm);

        /// <summary>Down-weight nearby spectra by retention-time distance (only used when
        /// <see cref="InterpolateRetentionTime"/> is false).</summary>
        public bool ApplyWeighting { get; init; } = true;

        /// <summary>Interpolate each multiplexed-spectrum row to the demuxed spectrum's RT via
        /// cubic Hermite splines. Cpp default true.</summary>
        public bool InterpolateRetentionTime { get; init; } = true;
    }

    private const int OverlapRegionsInApprox = 7;
    private const int CyclesInBlock = 3;

    private readonly Params _params;
    private ISpectrumList? _spectrumList;
    private IPrecursorMaskCodec? _maskCodec;
    private List<int> _spectrumIndices = new();

    /// <summary>Constructs an overlap demultiplexer with the given parameters.</summary>
    public OverlapDemultiplexer(Params? p = null) => _params = p ?? new Params();

    /// <inheritdoc/>
    public IReadOnlyList<int> SpectrumIndices => _spectrumIndices;

    /// <inheritdoc/>
    public void Initialize(ISpectrumList spectrumList, IPrecursorMaskCodec maskCodec)
    {
        ArgumentNullException.ThrowIfNull(spectrumList);
        ArgumentNullException.ThrowIfNull(maskCodec);
        _spectrumList = spectrumList;
        _maskCodec = maskCodec;
    }

    /// <inheritdoc/>
    public void BuildDeconvBlock(int spectrumIndex, IReadOnlyList<int> muxIndices,
        out Matrix<double> masks, out Matrix<double> signal)
    {
        if (_spectrumList is null || _maskCodec is null)
            throw new InvalidOperationException(
                "BuildDeconvBlock() Null SpectrumList or IPrecursorMaskCodec; OverlapDemultiplexer may not have been initialized.");

        var deconvSpectrum = _spectrumList.GetSpectrum(spectrumIndex, getBinaryData: true);
        var mzArr = deconvSpectrum.GetMZArray() ?? throw new InvalidOperationException("Deconv spectrum has no m/z array");
        var peakExtractor = new SpectrumPeakExtractor(mzArr.Data, _params.MassError);

        int numMuxSpectra = OverlapRegionsInApprox;
        int numDemuxSpectra = OverlapRegionsInApprox;
        int numTransitions = mzArr.Data.Count;
        masks = DenseMatrix.Create(numMuxSpectra, numDemuxSpectra, 0);
        signal = DenseMatrix.Create(numMuxSpectra, numTransitions, 0);

        // Center-of-mass of the demuxed spectrum's window indices.
        var deconvIndices = new List<int>();
        _maskCodec.SpectrumToIndices(deconvSpectrum, deconvIndices);
        double centerOfDeconvIndices = deconvIndices.Average();

        // Choose a contiguous lower-bound demux-window slice of length OverlapRegionsInApprox
        // centered roughly on the demuxed spectrum's m/z position.
        int idealLowerMzBound = (int)System.Math.Round(centerOfDeconvIndices - OverlapRegionsInApprox / 2.0);
        int lowerMzBound = System.Math.Max(idealLowerMzBound, 0);
        lowerMzBound = System.Math.Min(lowerMzBound, _maskCodec.NumDemuxWindows - OverlapRegionsInApprox);

        // Score each candidate mux spectrum by how close its window center-of-mass is to ours.
        var demuxWindowDistances = new List<(double Distance, int ScanIndex)>();
        var offsetIndices = new List<int>();
        foreach (var scanIndex in muxIndices)
        {
            var offsetSpectrum = _spectrumList.GetSpectrum(scanIndex, getBinaryData: true);
            _maskCodec.SpectrumToIndices(offsetSpectrum, offsetIndices);
            double centerOfOffsetIndices = offsetIndices.Average();
            demuxWindowDistances.Add((centerOfOffsetIndices - centerOfDeconvIndices, scanIndex));
        }

        // Sort by absolute distance (cpp uses an explicit |left|<|right| with 1e-3 tie-breaker).
        demuxWindowDistances.Sort((a, b) =>
        {
            double aa = System.Math.Abs(a.Distance);
            double bb = System.Math.Abs(b.Distance);
            if (System.Math.Abs(aa - bb) <= 1e-3) return 0;
            return aa < bb ? -1 : 1;
        });

        // Take the closest OverlapRegionsInApprox, then resort by signed distance for predictable
        // row order in the block.
        var bestMaskAverages = demuxWindowDistances.Take(OverlapRegionsInApprox).ToList();
        bestMaskAverages.Sort((a, b) =>
        {
            if (System.Math.Abs(a.Distance - b.Distance) <= 1e-3) return 0;
            return a.Distance < b.Distance ? -1 : 1;
        });
        var scansInDeconv = bestMaskAverages.Select(p => p.ScanIndex).ToList();

        // Fill the masks matrix from each chosen mux spectrum's m/z slice.
        for (int row = 0; row < scansInDeconv.Count; row++)
        {
            var muxSpectrum = _spectrumList.GetSpectrum(scansInDeconv[row], getBinaryData: true);
            var fullMaskRow = _maskCodec.GetMask(muxSpectrum);
            for (int col = 0; col < numDemuxSpectra; col++)
                masks[row, col] = fullMaskRow[lowerMzBound + col];
        }

        // Fill the signal matrix.
        if (_params.InterpolateRetentionTime)
        {
            double deconvStartTime = TryGetStartTime(deconvSpectrum)
                ?? throw new InvalidOperationException(
                    "BuildDeconvBlock() MS2 spectrum missing scan_start_time required for interpolation.");

            // For each chosen mux spectrum, find CyclesInBlock cycles' worth of identically-isolated
            // spectra (stride = SpectraPerCycle), record their (RT, binned-intensities) pairs, and
            // interpolate to the demuxed spectrum's RT.
            int specPerCycle = _maskCodec.SpectraPerCycle;
            var binnedIntensitiesCache = new List<Matrix<double>>();
            var scanTimesCache = new List<double[]>();
            for (int row = 0; row < scansInDeconv.Count; row++)
            {
                var binMatrix = DenseMatrix.Create(CyclesInBlock, numTransitions, 0);
                var times = new double[CyclesInBlock];
                var interpolationIndices = new List<int>();
                if (!DemuxHelpers.FindNearbySpectra(interpolationIndices, _spectrumList,
                    scansInDeconv[row], CyclesInBlock, specPerCycle))
                    throw new InvalidOperationException(
                        "BuildDeconvBlock() Not enough spectra to interpolate for the overlap.");

                for (int i = 0; i < interpolationIndices.Count; i++)
                {
                    var s = _spectrumList.GetSpectrum(interpolationIndices[i], getBinaryData: true);
                    times[i] = TryGetStartTime(s)
                        ?? throw new InvalidOperationException(
                            "BuildDeconvBlock() MS2 spectrum missing scan_start_time required for interpolation.");
                    peakExtractor.Extract(s, binMatrix, i);
                }
                binnedIntensitiesCache.Add(binMatrix);
                scanTimesCache.Add(times);
            }

            for (int row = 0; row < OverlapRegionsInApprox; row++)
                InterpolateMuxRegion(signal, row, deconvStartTime,
                    binnedIntensitiesCache[row], scanTimesCache[row]);
        }
        else
        {
            int specPerCycle = _maskCodec.SpectraPerCycle;
            for (int row = 0; row < scansInDeconv.Count; row++)
            {
                var s = _spectrumList.GetSpectrum(scansInDeconv[row], getBinaryData: true);
                double weight = 1.0;
                if (_params.ApplyWeighting)
                {
                    int scanDiff = spectrumIndex - scansInDeconv[row];
                    double t = 5.0 * scanDiff / specPerCycle;
                    weight = 1.0 / (1.0 + t * t);
                }
                peakExtractor.Extract(s, signal, row, weight);
            }
        }

        // Cache the spectrum's demux-window indices (relative to the slice we picked above).
        _spectrumIndices = new List<int>();
        foreach (var demuxIndex in deconvIndices)
        {
            if (demuxIndex < lowerMzBound)
                throw new InvalidOperationException(
                    "BuildDeconvBlock() Demux index slipped below the chosen lower bound — should not happen.");
            _spectrumIndices.Add(demuxIndex - lowerMzBound);
        }
    }

    /// <inheritdoc/>
    public void GetMatrixBlockIndices(int indexToDemux, List<int> muxIndices, double demuxBlockExtra = 0.0)
    {
        if (_spectrumList is null || _maskCodec is null)
            throw new InvalidOperationException(
                "GetMatrixBlockIndices() Null SpectrumList or IPrecursorMaskCodec; OverlapDemultiplexer may not have been initialized.");

        demuxBlockExtra = System.Math.Max(0.0, demuxBlockExtra);
        int numSpectraToFind = _maskCodec.SpectraPerCycle +
            (int)System.Math.Round(demuxBlockExtra * _maskCodec.SpectraPerCycle);
        if (!DemuxHelpers.FindNearbySpectra(muxIndices, _spectrumList, indexToDemux, numSpectraToFind))
            throw new InvalidOperationException(
                "GetMatrixBlockIndices() Not enough spectra to demultiplex this block");
    }

    private static void InterpolateMuxRegion(Matrix<double> signal, int rowToFill,
        double timeToInterpolate, Matrix<double> intensities, double[] scanTimes)
    {
        int numTransitions = intensities.ColumnCount;
        for (int transition = 0; transition < numTransitions; transition++)
        {
            var values = new double[scanTimes.Length];
            for (int i = 0; i < values.Length; i++) values[i] = intensities[i, transition];
            double interp = InterpolateOne(timeToInterpolate, scanTimes, values);
            signal[rowToFill, transition] = System.Math.Max(0.0, interp);
        }
    }

    private static double InterpolateOne(double pointToInterpolate, double[] points, double[] values)
    {
        // Cpp's "CubicHermiteSpline" wraps James Bremner's CSpline library, which is actually a
        // NATURAL cubic spline (zero 2nd derivative at endpoints), not Hermite — MathNet's
        // InterpolateNatural is the matching algorithm. Both expect strictly-increasing points.
        var spline = CubicSpline.InterpolateNatural(points, values);
        return spline.Interpolate(pointToInterpolate);
    }

    private static double? TryGetStartTime(Spectrum s)
    {
        if (s.ScanList.Scans.Count == 0) return null;
        var p = s.ScanList.Scans[0].Params.CvParam(CVID.MS_scan_start_time);
        if (p.Cvid == CVID.CVID_Unknown) return null;
        return p.ValueAs<double>();
    }
}
